internal/cmd/ping.go

```diff
@@ -0,0 +1,150 @@
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/fatih/color"
+	"github.com/pion/webrtc/v3"
+	"github.com/spf13/cobra"
+	"nhooyr.io/websocket"
+
+	"cdr.dev/coder-cli/coder-sdk"
+	"cdr.dev/coder-cli/pkg/clog"
+	"cdr.dev/coder-cli/wsnet"
+)
+
+func pingCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "ping [workspace_name]",
+		Short:   "Ping a Coder workspace",
+		Example: `coder ping my-dev`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			client, err := newClient(ctx, true)
+			if err != nil {
+				return err
+			}
+			workspace, err := findWorkspace(ctx, client, args[0], coder.Me)
+			if err != nil {
+				return err
+			}
+			if workspace.LatestStat.ContainerStatus != coder.WorkspaceOn {
+				return clog.Error("workspace not available",
+					fmt.Sprintf("current status: \"%s\"", workspace.LatestStat.ContainerStatus),
+					clog.BlankLine,
+					clog.Tipf("use \"coder workspaces rebuild %s\" to rebuild this workspace", workspace.Name),
+				)
+			}
+			servers, err := client.ICEServers(ctx)
+			if err != nil {
+				return err
+			}
+			url := client.BaseURL()
+			connectionStart := time.Now()
+			dialer, err := wsnet.DialWebsocket(ctx, wsnet.ConnectEndpoint(&url, workspace.ID, client.Token()), &wsnet.DialOptions{
+				ICEServers:         servers,
+				TURNProxyAuthToken: client.Token(),
+				TURNRemoteProxyURL: &url,
+				TURNLocalProxyURL:  &url,
+			}, &websocket.DialOptions{})
+			if err != nil {
+				return err
+			}
+			connectionMS := float64(time.Since(connectionStart).Microseconds()) / 1000
+			candidates, err := dialer.Candidates()
+			if err != nil {
+				return err
+			}
+			relay := candidates.Local.Typ == webrtc.ICECandidateTypeRelay
+			tunneled := false
+			properties := []string{}
+			candidateURLs := []string{}
+
+			for _, server := range servers {
+				if server.Username == wsnet.TURNProxyICECandidate().Username {
+					candidateURLs = append(candidateURLs, fmt.Sprintf("turns:%s", url.Host))
+					if !relay {
+						continue
+					}
+					tunneled = true
+					continue
+				}
+
+				candidateURLs = append(candidateURLs, server.URLs...)
+			}
+			properties = append(properties, fmt.Sprintf("candidates=%s", strings.Join(candidateURLs, ",")))
+
+			connectionText := "direct via STUN"
+			if relay {
+				connectionText = "proxied via TURN"
+			}
+			if tunneled {
+				connectionText = fmt.Sprintf("proxied via %s", url.Host)
+			}
+
+			fmt.Printf("%s %s %s (%s) %s\n",
+				color.New(color.Bold, color.FgWhite).Sprint("PING"),
+				workspace.Name,
+				color.New(color.Bold, color.FgGreen).Sprintf("connected in %.2fms", connectionMS),
+				connectionText,
+				strings.Join(properties, " "),
+			)
+
+			ticker := time.NewTicker(time.Second)
+			seq := 1
+			for {
+				select {
+				case <-ticker.C:
+					pingMS, connectionText, err := ping(ctx, client, dialer, tunneled, args[0])
+					if err != nil {
+						return err
+					}
+
+					fmt.Printf("%.2fms (%s) seq=%d\n",
```

Comment:
not needed for an MVP, of course, but would be nice if this printed stats when the user quit, like the actual ping command:

```
$ ping -c 3 google.com
PING google.com(sfo03s24-in-x0e.1e100.net (2607:f8b0:4005:80c::200e)) 56 data bytes
64 bytes from sfo03s24-in-x0e.1e100.net (2607:f8b0:4005:80c::200e): icmp_seq=1 ttl=118 time=10.8 ms
64 bytes from sfo03s24-in-x0e.1e100.net (2607:f8b0:4005:80c::200e): icmp_seq=2 ttl=118 time=10.3 ms
64 bytes from sfo03s24-in-x0e.1e100.net (2607:f8b0:4005:80c::200e): icmp_seq=3 ttl=118 time=11.2 ms

--- google.com ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2003ms
rtt min/avg/max/mdev = 10.329/10.785/11.239/0.371 ms
```

also, what happens if we lose a ping response? this seems to assume the ping packets always come back, but that seems difficult to guarantee in general on a lossy network?

Reply:
I believe the `Ping` func would return an error and we'd exit... I'll check.

Reply:
Fixed in the latest iteration. Just like ping, failures will display inline.

We don't have stats yet, but that can be added later on!